lib/partition_setup.pm

```diff
@@ -438,37 +438,29 @@ sub addboot {
     }
 }
 
-=head2 skip_select_first_hard_disk
-
- skip_select_first_hard_disk();
-
-Skip selecting first hard disk and return 0.
-Return 1 if no selection of hard disk is required or the first hard disk is already pre-selected.
-
-=cut
-sub skip_select_first_hard_disk {
-    return 1 if match_has_tag 'existing-partitions';          # no selection of hard-disk is required
-    if (match_has_tag('select-hard-disks-one-selected')) {    # first hd is already pre-selected
-        send_key $cmd{next};
-        return 1;
-    }
-    return 0;
-}
-
 =head2 select_first_hard_disk
 
  select_first_hard_disk();
 
 Select the first hard disk.
 
+The device should be [sv]da, other devices will be unselected. [sv]da device will also be
+force-selected at the end if needed (in some cases [sv]da is at the end of the list).
+
 =cut
 sub select_first_hard_disk {
-    my @tags = qw(existing-partitions hard-disk-dev-sdb-selected hard-disk-dev-sdc-selected hard-disk-dev-sdd-selected hard-disk-dev-non-sda-selected select-hard-disks-one-selected);
-    my $matched_needle = assert_screen \@tags;
-    return if skip_select_first_hard_disk;
+    # Try to handle most of the device type
+    my @tags    = 'existing-partitions';
+    my @devices = ('sdb' .. 'sdz', 'vdb' .. 'vdz', 'pmem0' .. 'pmem9');
+    foreach my $device (@devices) {
+        push @tags, "hard-disk-dev-$device-selected";
+    }
+    my $matched_needle = check_screen \@tags;           # save detected needle
+    return 1 if match_has_tag 'existing-partitions';    # no selection of hard-disk is required
+
     # SUT may have any number disks, only keep the first, unselect all other disks
     # In text mode, the needle has tag 'hard-disk-dev-non-sda-selected' and multiple hotkey_? tags as hints for unselecting
-    if (check_var('VIDEOMODE', 'text') && match_has_tag('hard-disk-dev-non-sda-selected')) {
+    if (check_var('VIDEOMODE', 'text') && check_screen('hard-disk-dev-non-sda-selected')) {
         foreach my $tag (@{$matched_needle->{needle}->{tags}}) {
```

Comment:
`$matched_needle` here contains the result from `check_screen()` in L458, but there is a second call to `check_screen()` in if conditional. Is this expected?

Reply:
Yes it is expected: the first `check_screen()` call doesn't check for `hard-disk-dev-non-sda-selected` tag, otherwise this tag is found pretty much all time, which lead to the issue we currently have. Instead, the new code check for all tags but this one, but it is still needed to check it at some point. This is why I have to do screen checking only with is needle.

Reply:
> Yes it is expected: the first `check_screen()` call doesn't check for `hard-disk-dev-non-sda-selected` tag, otherwise this tag is found pretty much all time, which lead to the issue we currently have. Instead, the new code check for all tags but this one, but it is still needed to check it at some point. This is why I have to do screen checking only with is needle.

Sure, but the issue has to do with `$matched_needle`. In the if-block, there is a for-loop going through the `hotkey_` tags  found on `$matched_needle` in order to deselect non-sda disks that were selected automatically. These `hotkey_` tags are only found on on `hard-disk-dev-non-sda-selected` needles in `products/sle/needles` (didn't find any in `products/opensuse/needles`). In the old code this was accomplished as `hard-disk-dev-non-sda-selected` was in the list of asserted needles at the start of the method and the if-block used `match_has_tag()`; but in the new code, `hard-disk-dev-non-sda-selected` is not used in the `check_screen()` used in the `$matched_needle` definition, so even if we get into the if-block in L463 due to `check_screen('hard-disk-dev-non-sda-selected')`, the matched needle from L458 is not the one that has the `hotkey_` tags. Basically when `VIDEOMODE=text`, the new code would not deselect non-sda disks.

Reply:
Your're right. I will have to test a screen textmode. In fact I should try to do lot of regression tests to be sure that the new code still work.

Reply:
I did a change to only check tags found in `hard-disk-dev-non-sda-selected` needle.